source/dub/generators/build.d

```diff
@@ -450,39 +457,103 @@ class BuildGenerator : ProjectGenerator {
-			if (buildsettings.workingDirectory.length) {
-				runcwd = Path(buildsettings.workingDirectory);
-				if (!runcwd.absolute) runcwd = cwd ~ runcwd;
-				logDiagnostic("Switching to %s", runcwd.toNativeString());
-				chdir(runcwd.toNativeString());
-			}
-			scope(exit) chdir(cwd.toNativeString());
-			if (!exe_file_path.absolute) exe_file_path = cwd ~ exe_file_path;
-			auto exe_path_string = exe_file_path.relativeTo(runcwd).toNativeString();
-			version (Posix) {
-				if (!exe_path_string.startsWith(".") && !exe_path_string.startsWith("/"))
-					exe_path_string = "./" ~ exe_path_string;
-			}
-			version (Windows) {
-				if (!exe_path_string.startsWith(".") && (exe_path_string.length < 2 || exe_path_string[1] != ':'))
-					exe_path_string = ".\\" ~ exe_path_string;
+		if (!workDir.length)
+			return cwd;
+		auto res = Path(workDir);
+		if (!res.absolute) res = cwd ~ res;
+		logDiagnostic("Switching to %s", res.toNativeString());
+		chdir(res.toNativeString());
+		return res;
+	}
+
+	private string[] runCmd(Path exe_file_path, Path cwd, Path runcwd, string[] run_args)
+	{
+		if (!exe_file_path.absolute) exe_file_path = cwd ~ exe_file_path;
+		auto exe_path_string = exe_file_path.relativeTo(runcwd).toNativeString();
+		version (Posix) {
+			if (!exe_path_string.startsWith(".") && !exe_path_string.startsWith("/"))
+				exe_path_string = "./" ~ exe_path_string;
+		}
+		version (Windows) {
+			if (!exe_path_string.startsWith(".") && (exe_path_string.length < 2 || exe_path_string[1] != ':'))
+				exe_path_string = ".\\" ~ exe_path_string;
+		}
+		return exe_path_string ~ run_args;
+	}
+
+	private void runTarget(Path exe_file_path, in BuildSettings buildsettings, string[] run_args, GeneratorSettings settings)
+	{
+		auto cwd = Path(getcwd());
+		auto runcwd = switchWorkDir(cwd, buildsettings.workingDirectory);
+		scope(exit) chdir(cwd.toNativeString());
+
+		auto cmd = runCmd(exe_file_path, cwd, runcwd, run_args);
+		logInfo("Running '%-(%s, %)'", cmd);
+
+		if (settings.runCallback) {
+			auto res = execute(cmd);
+			settings.runCallback(res.status, res.output);
+		} else {
+			auto prg_pid = spawnProcess(cmd);
+			auto result = prg_pid.wait();
+			enforce(result == 0, "Program exited with code "~to!string(result));
+		}
+	}
+
+	private void watchTarget(Path exe_file_path, in BuildSettings buildsettings, string[] run_args, GeneratorSettings settings, in TargetInfo[string] targets)
+	{
+		Watcher w;
+		logInfo("Setting up file watches.");
+		foreach (_, tgt; targets)
+		{
+			auto bs = tgt.buildSettings;
+			import std.range : chain;
+			foreach (file; chain(bs.sourceFiles, bs.importFiles, bs.stringImportFiles))
+				w.addFile(file);
+		}
+
+		auto cwd = Path(getcwd());
+		auto runcwd = switchWorkDir(cwd, buildsettings.workingDirectory);
+		scope(exit) chdir(cwd.toNativeString());
+
+		auto cmd = runCmd(exe_file_path, cwd, runcwd, run_args);
+		assert(!settings.runCallback);
+
+		while (true)
+		{
+			logInfo("Running '%-(%s, %)'", cmd);
```

Comment:
Interesting, never seen that format syntax before. Shouldn't it be `%-(%s %)` (without the comma), though?